### PR TITLE
Exclude only named children without injection.include-children

### PR DESCRIPTION
**Pull request by MDeiml**
This fixes problems arising in #3108.

The markdown parser needs to exclude children for language injection, but only named ones. This is *different* from the behavior in tree-sitter-highlight, but to me it seems more sensible. 

To give a technical example: Otherwise there is almost no way to write something like
```js
rule: $ => seq('(', repeat(choice($._content, $.excluded)), ')'),
```
where `$._content` and the parentheses should be included and `$.excluded` excluded. This is because there will always be `'('` and `')'` nodes as a child of `rule`.

### CODE REVIEW

**Comment by the-mikedavis** [comment]
If you want a node from the tree to be queryable, it should be named: that's the point of named nodes. I think you should restructure the grammar to expose the nodes you want to capture in queries.

I can't test this at the moment but I suspect this breaks injections for some templating languages or maybe Rust macros? I'll give it a look when I have a chance

**Comment by MDeiml** [comment]
> If you want a node from the tree to be queryable, it should be named: that's the point of named nodes. I think you should restructure the grammar to expose the nodes you want to capture in queries.

That's my point. I expose `inline` nodes and `block_continuation` nodes. I want to capture the former while excluding the latter. I don't want to exclude unnamed nodes that I have less control over when writing the grammar.

Also notice how not excluding unnamed nodes is strictly more expressive. In a tree-sitter grammar an unnamed node can always be turned into a named one with `alias(...)`. I think this is exactly what you meant?

> I can't test this at the moment but I suspect this breaks injections for some templating languages or maybe Rust macros? I'll give it a look when I have a chance

I'm gonna have a look myself too. If that's the case and the queries can't easily be changed to work with the new behavior: how do you feel about adding a `include-unnamed` or `exclude-named` flag? I'm not sure it's possible to make my grammar work without this.

**Comment by MDeiml** [comment]
It seems you're right. Many languages rely on excluding unnamed nodes for injecting the `comment` language.

**Comment by the-mikedavis** [comment]
A separate property like include/exclude-unnamed sounds like a good addition to me 👍

File: helix-core/src/syntax.rs [hunk 1]
```diff
@@ -1808,6 +1825,7 @@ fn injection_for_match<'a>(
     }
 
     let mut include_children = false;
+    let mut include_unnamed_children = false;
```
Comment (the-mikedavis) [inline helix-core/src/syntax.rs:1828]: Instead of two booleans, I think it would be cleaner to use an enum

Comment (MDeiml) [reply helix-core/src/syntax.rs]: I think a enum with tree members makes the most sense here :+1:

**Review by the-mikedavis — APPROVED** [review]
This looks good, thanks for working on it 👍

**Pull request merged**